File: 10-edge/02-continuous-batching-and-paged-kv/java/CancelPropagation.java
// Layer 10 - Topic 2: does hanging up actually free the KV blocks? (Java)
//
// What this demonstrates
//     The same experiment as the other five, on Java 21+ virtual threads --
//     the workload Loom was designed for: many concurrent requests, each
//     mostly waiting. A stub model server streams 40 tokens at 100ms each
//     while watching for its caller to leave; a gateway sits in front with
//     two handlers; a client hangs up after 500ms against each.
//
//       /naive       one virtual thread per request, blocking style, reads
//                    the upstream to EOF and then writes it downstream.
//                    Nothing is watching the client socket, so the upstream
//                    generation finishes for a response that is discarded.
//       /cancelling  a second virtual thread blocks on a read of the CLIENT
//                    socket. A read returning -1 is the hang-up, and it
//                    calls interrupt() on the worker.
//
// The Java-specific finding this file exists to check
//     Thread.interrupt() has historically been useless against a blocking
//     socket read: java.net.Socket streams are not interruptible channels,
//     so the interrupt sets a flag nobody looks at until the read returns
//     on its own. Under Loom, socket I/O on a VIRTUAL thread is implemented
//     on top of NIO and the platform carrier is never actually blocked --
//     so interrupt() unparks the virtual thread and the read throws. The
//     table below is the check, not the claim: run it and see which column
//     the /cancelling row lands in.
//
// What to look for
//     - /naive: upstream decodes all 40 tokens, ~3.5s of it after the
//       client stopped listening.
//     - /cancelling: how quickly the upstream sees EOF, and whether it sees
//       it at all. If it does not, interrupt() did not reach the read and
//       you would need to close the upstream socket from the watcher --
//       the C++ answer, in Java.
//
// No dependencies. Binds 127.0.0.1 only. Runs with no arguments:
//     cd java && javac CancelPropagation.java -d /tmp/javabuild \
//       && java -cp /tmp/javabuild CancelPropagation

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class CancelPropagation {
    static final int TOKENS = 40;
    static final long TOKEN_INTERVAL_MS = 100;        // 4.0s of "decode"
    static final long CLIENT_HANGS_UP_AFTER_MS = 500;

    record Observation(boolean aborted, int tokens, double seconds) {}

    static final List<Observation> LEDGER = new ArrayList<>();

    static void record(Observation o) {
        synchronized (LEDGER) {
            LEDGER.add(o);
        }
    }

    // ---------------------------------------------------------------------
    // The stub model server. A read returning -1 on its own socket is the
    // only signal it gets that the consumer of this generation has gone.
    // ---------------------------------------------------------------------
    static void upstreamConnection(Socket sock) {
        long start = System.nanoTime();
        int sent = 0;
        boolean aborted = false;
        try (sock) {
            InputStream in = sock.getInputStream();
            OutputStream out = sock.getOutputStream();
            readRequestHead(in);
            out.write("HTTP/1.1 200 OK\r\nContent-Type: text/event-stream\r\n\r\n"
                    .getBytes(StandardCharsets.US_ASCII));
            out.flush();

            // A virtual thread parked on read() costs nothing, which is what
            // makes "one watcher per in-flight request" a reasonable design
            // rather than a thread-count disaster.
            Thread peerWatcher = Thread.ofVirtual().start(() -> {
                try {
                    if (in.read() == -1) sock.close();
                } catch (IOException ignored) {
                    // socket closed under us: same conclusion
                }
            });

            for (int i = 0; i < TOKENS; i++) {
                if (sock.isClosed()) { aborted = true; break; }
                try {
                    out.write(("data: token " + i + "\n\n").getBytes(StandardCharsets.US_ASCII));
                    out.flush();
                } catch (IOException e) {
                    aborted = true;
                    break;
                }
                sent = i + 1;
                Thread.sleep(TOKEN_INTERVAL_MS);
            }
            peerWatcher.interrupt();
        } catch (IOException | InterruptedException e) {
            aborted = true;
        } finally {
            record(new Observation(aborted, sent, (System.nanoTime() - start) / 1e9));
        }
    }

    // ---------------------------------------------------------------------
    // The gateway.
    // ---------------------------------------------------------------------
    static void gatewayConnection(Socket client, int upstreamPort) {
        try (client) {
            String head = readRequestHead(client.getInputStream());
            boolean cancelling = head.contains("/cancelling");

            Socket upstream = new Socket(InetAddress.getLoopbackAddress(), upstreamPort);
            Thread worker = Thread.currentThread();
            Thread watcher = null;

            if (cancelling) {
                InputStream fromClient = client.getInputStream();
                watcher = Thread.ofVirtual().start(() -> {
                    try {
                        if (fromClient.read() == -1) {
                            // The whole fix. On a virtual thread the worker
                            // is parked on NIO, not on a carrier, so this
                            // interrupt reaches the blocked read.
                            worker.interrupt();
                        }
                    } catch (IOException ignored) {
                        worker.interrupt();
                    }
                });
            }

            try (upstream) {
                OutputStream up = upstream.getOutputStream();
                up.write(("POST /completions HTTP/1.1\r\nHost: localhost\r\n"
                        + "Content-Length: 0\r\n\r\n").getBytes(StandardCharsets.US_ASCII));
                up.flush();

                // Buffer the upstream response, then reply. Deliberate: a
                // streaming forward would fail its first downstream write
                // and tear the upstream down that way, which is a second
                // safety net that would hide whether interrupt() worked.
                InputStream from = upstream.getInputStream();
                byte[] buf = new byte[4096];
                var body = new java.io.ByteArrayOutputStream();
                int n;
                while ((n = from.read(buf)) > 0) body.write(buf, 0, n);
                client.getOutputStream().write(body.toByteArray());
            } catch (IOException e) {
                // Interrupting a virtual thread blocked in socket I/O closes
                // the channel and surfaces here.
            } finally {
                if (watcher != null) watcher.interrupt();
                Thread.interrupted();  // clear the flag before the thread is reused
            }
        } catch (IOException ignored) {
        }
    }

    static String readRequestHead(InputStream in) throws IOException {
        byte[] buf = new byte[2048];
        int n = in.read(buf);
        return n <= 0 ? "" : new String(buf, 0, n, StandardCharsets.US_ASCII);
    }

    // A raw socket, on purpose: an HTTP client's timeout returns control to
    // your code without necessarily closing the TCP connection, so the
    // server would see nothing and the experiment would measure something
    // else entirely.
    static void hangUpOn(int gatewayPort, String path) throws Exception {
        Socket sock = new Socket(InetAddress.getLoopbackAddress(), gatewayPort);
        sock.getOutputStream().write(("POST " + path + " HTTP/1.1\r\nHost: localhost\r\n"
                + "Content-Length: 0\r\n\r\n").getBytes(StandardCharsets.US_ASCII));
        sock.getOutputStream().flush();
        sock.setSoTimeout((int) CLIENT_HANGS_UP_AFTER_MS);
        try {
            byte[] buf = new byte[4096];
            while (sock.getInputStream().read(buf) > 0) { /* read until we give up */ }
        } catch (IOException expected) {
            // the read timeout: this is the moment the user closes the tab
        }
        sock.close();  // the hang-up
    }

    public static void main(String[] args) throws Exception {
        ExecutorService vthreads = Executors.newVirtualThreadPerTaskExecutor();

        ServerSocket upstreamLn = new ServerSocket(0, 16, InetAddress.getLoopbackAddress());
        vthreads.submit(() -> {
            while (!upstreamLn.isClosed()) {
                Socket s = upstreamLn.accept();
                vthreads.submit(() -> upstreamConnection(s));
            }
            return null;
        });

        ServerSocket gatewayLn = new ServerSocket(0, 16, InetAddress.getLoopbackAddress());
        vthreads.submit(() -> {
            while (!gatewayLn.isClosed()) {
                Socket s = gatewayLn.accept();
                vthreads.submit(() -> gatewayConnection(s, upstreamLn.getLocalPort()));
            }
            return null;
        });

        System.out.println("Java " + Runtime.version().feature()
                + " / virtual threads - cancellation on client disconnect");
        System.out.printf("  upstream streams %d tokens x %dms = %.1fs of decode%n",
                TOKENS, TOKEN_INTERVAL_MS, TOKENS * TOKEN_INTERVAL_MS / 1000.0);
        System.out.printf("  client hangs up after %.1fs%n%n",
                CLIENT_HANGS_UP_AFTER_MS / 1000.0);
        System.out.printf("  %-14s %-16s %14s %13s %8s%n",
                "handler", "upstream saw", "tokens decoded", "upstream ran", "wasted");
        System.out.println("  " + "-".repeat(70));

        for (String path : new String[] {"/naive", "/cancelling"}) {
            synchronized (LEDGER) { LEDGER.clear(); }
            hangUpOn(gatewayLn.getLocalPort(), path);

            long deadline = System.nanoTime()
                    + TimeUnit.MILLISECONDS.toNanos(TOKENS * TOKEN_INTERVAL_MS + 1000);
            Observation obs = new Observation(false, 0, Double.NaN);
            while (System.nanoTime() < deadline) {
                synchronized (LEDGER) {
                    if (!LEDGER.isEmpty()) { obs = LEDGER.get(0); break; }
                }
                Thread.sleep(50);
            }

            double wasted = Math.max(0.0, obs.seconds() - CLIENT_HANGS_UP_AFTER_MS / 1000.0);
            System.out.printf("  %-14s %-16s %14d %12.2fs %7.2fs%n", path,
                    obs.aborted() ? "cancelled" : "nothing", obs.tokens(),
                    obs.seconds(), wasted);
        }

        System.out.println();
        System.out.println("  'wasted' is decode time spent on a response nobody read. On a");
        System.out.println("  loaded server those KV blocks stayed allocated the whole time,");
        System.out.println("  so the scheduler could not admit somebody who was still waiting.");
        System.out.println();
        System.out.println("  Watch the /cancelling row: it is a direct test of whether");
        System.out.println("  Thread.interrupt() reaches a socket read on a virtual thread.");

        upstreamLn.close();
        gatewayLn.close();
        vthreads.shutdownNow();
        System.exit(0);
    }
}

File: 10-edge/02-continuous-batching-and-paged-kv/cpp/cancel_propagation.cpp
// Layer 10 - Topic 2: does hanging up actually free the KV blocks? (C++)
//
// What this demonstrates
//     The same experiment as the other five, in the language that has no
//     cancellation mechanism at all -- which is exactly why it is here. A
//     thread blocked in read() on the upstream socket learns nothing about
//     the downstream one. Nobody is going to cancel anything for you, so
//     the fix has to be spelled out in full:
//
//       /naive       connect upstream, read it to EOF, write the result to
//                    the client. The thread is parked in read() the whole
//                    time. When the client leaves, this code has no way to
//                    find out and no way to react if it did.
//       /cancelling  a watcher thread poll()s the CLIENT socket for EOF,
//                    and on EOF calls shutdown(upstream_fd, SHUT_RDWR).
//                    That forces the blocked read() in the worker thread to
//                    return, and closes the upstream connection so the
//                    engine frees the sequence.
//
//     Every other runtime in this topic is doing some version of that
//     watcher for you: Go's net/http cancels r.Context(), Node fires
//     req.on('close'), tokio drops the losing future, Starlette throws into
//     the generator, the JVM interrupts the virtual thread. Here you can
//     see the machinery, because you have to write it.
//
// What to look for
//     - The fix is a thread and a shutdown() call, and nothing about the
//       "fast, compiled language" helped. This is the same argument as
//       Layer 1's C++ pairing: the failure mode is about scheduling and
//       ownership, not about speed.
//     - shutdown() rather than close(): close() on an fd another thread is
//       blocked in read() on is a use-after-free waiting to happen if that
//       fd number gets reused. shutdown() leaves the descriptor valid and
//       makes the pending read return 0.
//
// macOS note: this uses poll(2), which exists on Darwin. epoll(7) does not
// -- see Layer 1's portability notes. Nothing here needs an event loop
// anyway; the point is the absence of one.
//
// Build and run (no arguments, binds 127.0.0.1 only):
//     c++ -O2 -std=c++20 -pthread -o /tmp/cancel_cpp cpp/cancel_propagation.cpp \
//       && /tmp/cancel_cpp

#include <arpa/inet.h>
#include <netinet/in.h>
#include <poll.h>
#include <sys/socket.h>
#include <unistd.h>

#include <algorithm>
#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdio>
#include <cstring>
#include <mutex>
#include <string>
#include <thread>
#include <vector>

namespace {

constexpr int kTokens = 40;
constexpr auto kTokenInterval = std::chrono::milliseconds(100);  // 4.0s of decode
constexpr auto kClientHangsUpAfter = std::chrono::milliseconds(500);

using Clock = std::chrono::steady_clock;

struct Observation {
    bool aborted;
    int tokens;
    double seconds;
};

std::mutex g_ledger_mu;
std::vector<Observation> g_ledger;

void record(Observation o) {
    std::lock_guard<std::mutex> lock(g_ledger_mu);
    g_ledger.push_back(o);
}

int listen_on_ephemeral_port(uint16_t* port_out) {
    int fd = ::socket(AF_INET, SOCK_STREAM, 0);
    int one = 1;
    ::setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
    sockaddr_in addr{};
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    addr.sin_port = 0;
    ::bind(fd, reinterpret_cast<sockaddr*>(&addr), sizeof(addr));
    ::listen(fd, 16);
    socklen_t len = sizeof(addr);
    ::getsockname(fd, reinterpret_cast<sockaddr*>(&addr), &len);
    *port_out = ntohs(addr.sin_port);
    return fd;
}

int connect_to(uint16_t port) {
    int fd = ::socket(AF_INET, SOCK_STREAM, 0);
    sockaddr_in addr{};
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
    addr.sin_port = htons(port);
    if (::connect(fd, reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) != 0) {
        ::close(fd);
        return -1;
    }
    return fd;
}

// Has the peer closed its end? POLLIN with a zero-length peek is EOF.
// MSG_PEEK so this never consumes a byte the caller still wants.
bool peer_closed(int fd) {
    pollfd p{fd, POLLIN | POLLHUP, 0};
    if (::poll(&p, 1, 0) <= 0) return false;
    if (p.revents & (POLLHUP | POLLERR | POLLNVAL)) return true;
    char c;
    ssize_t n = ::recv(fd, &c, 1, MSG_PEEK | MSG_DONTWAIT);
    return n == 0;
}

std::string read_request_head(int fd) {
    char buf[2048];
    ssize_t n = ::recv(fd, buf, sizeof(buf) - 1, 0);
    if (n <= 0) return {};
    buf[n] = '\0';
    return std::string(buf, static_cast<size_t>(n));
}

// ---------------------------------------------------------------------------
// The stub model server.
// ---------------------------------------------------------------------------
void upstream_connection(int fd) {
    read_request_head(fd);
    const char* head =
        "HTTP/1.1 200 OK\r\nContent-Type: text/event-stream\r\n\r\n";
    ::send(fd, head, std::strlen(head), 0);

    const auto start = Clock::now();
    int sent = 0;
    bool aborted = false;
    for (int i = 0; i < kTokens; ++i) {
        if (peer_closed(fd)) {
            aborted = true;
            break;
        }
        std::string chunk = "data: token " + std::to_string(i) + "\n\n";
        // MSG_NOSIGNAL does not exist on Darwin, so SIGPIPE is ignored in
        // main() and this send() returns -1/EPIPE instead of killing us.
        if (::send(fd, chunk.data(), chunk.size(), 0) < 0) {
            aborted = true;
            break;
        }
        sent = i + 1;
        std::this_thread::sleep_for(kTokenInterval);
    }
    record({aborted, sent,
            std::chrono::duration<double>(Clock::now() - start).count()});
    ::close(fd);
}

// ---------------------------------------------------------------------------
// The gateway.
// ---------------------------------------------------------------------------
void gateway_connection(int client_fd, uint16_t upstream_port) {
    const std::string head = read_request_head(client_fd);
    const bool cancelling = head.find("/cancelling") != std::string::npos;

    int up_fd = connect_to(upstream_port);
    if (up_fd < 0) {
        ::close(client_fd);
        return;
    }
    const char* req =
        "POST /completions HTTP/1.1\r\nHost: localhost\r\nContent-Length: 0\r\n\r\n";
    ::send(up_fd, req, std::strlen(req), 0);

    std::atomic<bool> done{false};
    std::thread watcher;
    if (cancelling) {
        // The fix, in full. No runtime is going to do this for you.
        watcher = std::thread([client_fd, up_fd, &done] {
            while (!done.load(std::memory_order_relaxed)) {
                if (peer_closed(client_fd)) {
                    // shutdown(), not close(): the worker thread is blocked
                    // in recv() on this descriptor. shutdown() makes that
                    // recv return 0 while the descriptor stays valid, so
                    // there is no window for the fd number to be reused
                    // under the blocked thread.
                    ::shutdown(up_fd, SHUT_RDWR);
                    return;
                }
                std::this_thread::sleep_for(std::chrono::milliseconds(20));
            }
        });
    }

    // Read the upstream to EOF, buffering. This is the blocking call the
    // whole topic is about: without the watcher above, this thread cannot
    // learn anything about the client while it sits here.
    std::string body;
    char buf[4096];
    for (;;) {
        ssize_t n = ::recv(up_fd, buf, sizeof(buf), 0);
        if (n <= 0) break;
        body.append(buf, static_cast<size_t>(n));
    }
    done.store(true, std::memory_order_relaxed);
    if (watcher.joinable()) watcher.join();

    ::send(client_fd, body.data(), body.size(), 0);
    ::close(up_fd);
    ::close(client_fd);
}

void accept_loop(int listen_fd, uint16_t upstream_port, bool is_gateway) {
    for (;;) {
        int fd = ::accept(listen_fd, nullptr, nullptr);
        if (fd < 0) return;
        if (is_gateway) {
            std::thread(gateway_connection, fd, upstream_port).detach();
        } else {
            std::thread(upstream_connection, fd).detach();
        }
    }
}

// A raw socket, on purpose: a client library's timeout returns control to
// your code without necessarily closing the TCP connection, so the server
// would see nothing and the experiment would measure something else.
void hang_up_on(uint16_t gateway_port, const char* path) {
    int fd = connect_to(gateway_port);
    if (fd < 0) return;
    std::string req = std::string("POST ") + path +
                      " HTTP/1.1\r\nHost: localhost\r\nContent-Length: 0\r\n\r\n";
    ::send(fd, req.data(), req.size(), 0);
    const auto deadline = Clock::now() + kClientHangsUpAfter;
    char buf[4096];
    while (Clock::now() < deadline) {
        pollfd p{fd, POLLIN, 0};
        if (::poll(&p, 1, 20) > 0 && ::recv(fd, buf, sizeof(buf), 0) <= 0) break;
    }
    ::close(fd);  // the hang-up
}

}  // namespace

int main() {
    // A send() to a socket the peer closed raises SIGPIPE by default, which
    // kills the process. Ignore it and take the -1/EPIPE return instead.
    ::signal(SIGPIPE, SIG_IGN);

    uint16_t upstream_port = 0;
    uint16_t gateway_port = 0;
    int upstream_ln = listen_on_ephemeral_port(&upstream_port);
    int gateway_ln = listen_on_ephemeral_port(&gateway_port);
    std::thread(accept_loop, upstream_ln, upstream_port, false).detach();
    std::thread(accept_loop, gateway_ln, upstream_port, true).detach();

    std::printf("C++ / POSIX sockets - cancellation on client disconnect\n");
    std::printf("  upstream streams %d tokens x %lldms = %.1fs of decode\n", kTokens,
                static_cast<long long>(kTokenInterval.count()),
                kTokens * kTokenInterval.count() / 1000.0);
    std::printf("  client hangs up after %.1fs\n\n",
                kClientHangsUpAfter.count() / 1000.0);
    std::printf("  %-14s %-16s %14s %13s %8s\n", "handler", "upstream saw",
                "tokens decoded", "upstream ran", "wasted");
    std::printf("  ----------------------------------------------------------------------\n");

    for (const char* path : {"/naive", "/cancelling"}) {
        {
            std::lock_guard<std::mutex> lock(g_ledger_mu);
            g_ledger.clear();
        }
        hang_up_on(gateway_port, path);

        const auto deadline =
            Clock::now() + kTokenInterval * kTokens + std::chrono::seconds(1);
        Observation obs{false, 0, 0.0};
        for (;;) {
            {
                std::lock_guard<std::mutex> lock(g_ledger_mu);
                if (!g_ledger.empty()) {
                    obs = g_ledger.front();
                    break;
                }
            }
            if (Clock::now() > deadline) break;
            std::this_thread::sleep_for(std::chrono::milliseconds(50));
        }

        const double wasted =
            std::max(0.0, obs.seconds - kClientHangsUpAfter.count() / 1000.0);
        std::printf("  %-14s %-16s %14d %12.2fs %7.2fs\n", path,
                    obs.aborted ? "cancelled" : "nothing", obs.tokens, obs.seconds,
                    wasted);
    }

    std::printf("\n");
    std::printf("  'wasted' is decode time spent on a response nobody read. On a\n");
    std::printf("  loaded server those KV blocks stayed allocated the whole time,\n");
    std::printf("  so the scheduler could not admit somebody who was still waiting.\n\n");
    std::printf("  The fix cost one thread and one shutdown() call. Every other\n");
    std::printf("  runtime in this topic ships that watcher; here you write it.\n");
    return 0;
}
